db/config.cc

```diff
@@ -1146,15 +1146,25 @@ db::fs::path db::config::get_conf_sub(db::fs::path sub) {
 }
 
 bool db::config::check_experimental(experimental_features_t::feature f) const {
-    if (experimental()
-        && f != experimental_features_t::feature::UNUSED
-        && f != experimental_features_t::feature::CONSISTENT_TOPOLOGY_CHANGES
-        && f != experimental_features_t::feature::BROADCAST_TABLES
-        && f != experimental_features_t::feature::TABLETS) {
-            return true;
+    const auto features_disabled_by_default = {
+        experimental_features_t::feature::UNUSED,
+        experimental_features_t::feature::CONSISTENT_TOPOLOGY_CHANGES,
+        experimental_features_t::feature::BROADCAST_TABLES,
+        experimental_features_t::feature::TABLETS,
+    };
+    if (experimental() &&
+        std::ranges::none_of(features_disabled_by_default,
```

Comment:
!std::ranges::contains(features_disabled_by_default, f)?

Also, why call it features_disabled_by_default? they're just experimental, no?

Reply:
> !std::ranges::contains(features_disabled_by_default, f)?

unfortunately, `std::ranges::contains()` is part of C++23. in libstdc++, this function is rightfully guarded by 
```c++
#if __cplusplus > 202002L
```
and we are using `--std=gnu++20` at this moment.

> 
> Also, why call it features_disabled_by_default? they're just experimental, no?

my guess is that, the authors of these features don't want to enable them by default even if `--experimental=true`. so in this context, i think "default" means enabled `experimental=true`.  i am quoting the commit message of one of the changes adding a feature to the list:

>    db: add experimental option for raft
>     
>     Introduce `raft` experimental option.
>     Adjust the tests accordingly to accomodate the new option.
>     
>     It's not enabled by default when providing
>     `--experimental=true` config option and should be
>     requested explicitly via `--experimental-options=raft`
>     config option.

see 22794efc220ef85e01f343b255b603a72e0de1f9

probably i should rename it to something more obvious, like `highly_experimental_features` ?

Reply:
Traditionally, we only adopt a standard after it's no longer a draft. This way, if we start to use a C++23 feature and it's ejected from C++23, we don't end up with code that doesn't match any released version.

It's very unlikely anything will change from the draft, and even if it does, it will take compilers time to adjust (though I guess they'll only advertise -std=c++23 after fixing up stuff). But maybe we can relax the rule for this repository.

Reply:
I don't understand highly_experimental_features either. What's the difference compared to other experimental features?

wrt --experimental=true, I thought we deprecated it.

Reply:
> Traditionally, we only adopt a standard after it's no longer a draft. This way, if we start to use a C++23 feature and it's ejected from C++23, we don't end up with code that doesn't match any released version.

C++23 is no longer a draft. see https://isocpp.org/std/status. and the committee has started working on C++26, see https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p1000r5.pdf 

> 
> It's very unlikely anything will change from the draft, and even if it does, it will take compilers time to adjust (though I guess they'll only advertise -std=c++23 after fixing up stuff). But maybe we can relax the rule for this repository.

Reply:
> I don't understand highly_experimental_features either. What's the difference compared to other experimental features?
> 
> wrt --experimental=true, I thought we deprecated it.

it was marked deprecated in 8b917f7c99c6459d252fd848685f6d54b295e6ae . instead of asking the authors of these highly experimental features for the difference in comparison with other experimental features which can be enabled by this option, i think the better way forward is to remove this option.

i will update dtest first.

Reply:
> or, we can start using C++23? as quite a few C++23 features are already [supported](https://en.cppreference.com/w/cpp/23) by the compiler and the standard library.

a blocker might be https://github.com/scylladb/seastar/issues/1772 . i think the new tuple-alike constructor introduced in C++23 also contributes to this failure.